resolwe/flow/managers/base.py

```diff
@@ -577,7 +577,74 @@ def communicate(self, run_sync=False, verbosity=1, save_settings=True):
+    def _data_execute(self, data, program, executor, verbosity):
+        """Execute the Data object.
+
+        The activities carried out here include target directory
+        preparation, executor copying, setting serialization and actual
+        execution of the object.
+
+        :param data: The :class:`~resolwe.flow.models.Data` object to
+            execute.
+        :param program: The process text the manager got out of
+            execution engine evaluation.
+        :param executor: The executor to use for this object.
+        :param verbosity: The logging verbosity with which to execute
+            the object.
+        """
+        if not program:
+            return
+
+        logger.debug(__(
+            "Manager '{}.{}' preparing Data object {} for processing.",
+            self.__class__.__module__,
+            self.__class__.__name__,
+            data.id
+        ))
+
+        # Prepare the executor's environment.
+        try:
+            priority = 'normal'
+            if data.process.persistence == Process.PERSISTENCE_TEMP:
+                # TODO: This should probably be removed.
+                priority = 'high'
+            if data.process.scheduling_class == Process.SCHEDULING_CLASS_INTERACTIVE:
+                priority = 'high'
+
+            program = self._include_environment_variables(program)
+
+            data_dir = self._prepare_data_dir(data.id)
+            executor_module, runtime_dir = self._prepare_executor(data.id, executor)
+            self._prepare_context(data.id, data_dir, runtime_dir, verbosity=verbosity)
+            self._prepare_script(runtime_dir, program)
+            argv = [
+                '/bin/bash',
+                '-c',
+                self.settings_actual.get('FLOW_EXECUTOR', {}).get('PYTHON', '/usr/bin/env python') +
+                ' -m executors ' + executor_module
+            ]
+        except PermissionDenied as error:
+            data.status = Data.STATUS_ERROR
+            data.process_error.append("Permission denied for process: {}".format(error))
+            data.save()
+            return
+        except OSError as err:
+            logger.error(__(
+                "OSError occurred while preparing data {} (will skip): {}",
+                data.id, err
+            ))
+            return
+
+        # Hand off to the run() method for execution.
+        if verbosity >= 1:
+            print("Running", runtime_dir)
+        self.state.executor_count.add(1)
+        # Set execution claim on the semaphore.
+        new_sema = self.state.sync_semaphore.add(1)
+        logger.debug(__("Manager changed sync_semaphore UP to {} on executor start.", new_sema))
+        self.run(data.id, runtime_dir, argv, priority=priority, verbosity=verbosity)
+
+    def _data_scan(self, run_sync=False, verbosity=1, executor='resolwe.flow.executors.local', **kwargs):
         """Resolve task dependencies and run the task.
```

Comment:
Would `Resolwe task dependencies, mark Data objects for execution.` be more appropriate after this refactoring?

Reply:
Not entirely, since the functionality hasn't changed in any way. Objects aren't marked for anything either, they're handed directly to the runner function.

Reply:
I want the docstring to reflect the change of splitting the execution part to the `_data_exection()` method.
What about `Resolve task dependencies and trigger task execution` since we use Django's signals to trigger execution after transaction is commited?

Reply:
We don't use signals for this. The behaviour of these two functions is precisely the same as it was when they were part of `_communicate`.

Reply:
@jberci, thanks for the explanation, I agree with your argument.